ouroboros-network/src/Ouroboros/Network/PeerSelection/EstablishedPeers.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE DeriveFunctor  #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Ouroboros.Network.PeerSelection.EstablishedPeers
+  ( EstablishedPeers
+  , establishedReady
+  , establishedStatus
+  , empty
+  , toMap
+
+  , size
+
+  , member
+
+  , insert
+  , delete
+  , deletePeers
+
+  , updateStatus
+  , updateStatuses
+
+  , setCurrentTime
+  , minActivateTime
+  , setActivateTime
+
+  , invariant
+  ) where
+
+import           Prelude
+
+import           Data.Foldable (foldl')
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.OrdPSQ (OrdPSQ)
+import qualified Data.OrdPSQ as PSQ
+import qualified Data.Set as Set
+import           Data.Set (Set)
+
+import           Control.Monad.Class.MonadTime
+import           Control.Exception (assert)
+
+import           Ouroboros.Network.PeerSelection.Types
+
+
+data EstablishedPeers peeraddr peerconn = EstablishedPeers {
+    -- | Peers which are either ready do become active or are active.
+    --
+    establishedReady  :: !(Map peeraddr peerconn),
+
+    -- | 'PeerStatus' of all established peers.
+    establishedStatus :: !(Map peeraddr PeerStatus),
+
+    -- | Peers which are not ready to become active.
+    nextActivateTimes :: !(OrdPSQ peeraddr Time peerconn)
+  }
+  deriving (Show, Functor)
+
+
+empty :: EstablishedPeers peeraddr perconn
+empty = EstablishedPeers Map.empty Map.empty PSQ.empty
+
+
+invariant :: Ord peeraddr
+          => EstablishedPeers peeraddr peerconn
+          -> Bool
+invariant EstablishedPeers { establishedReady,
+                             establishedStatus,
+                             nextActivateTimes } =
+     -- established connections and nextActivateTimes do not intersect
+     Set.null
+       (Set.intersection
+         (Map.keysSet establishedReady)
+         (Set.fromList (PSQ.keys nextActivateTimes)))
+
+     -- established connection + nextActivate times has the same keys as
+     -- established status
+  &&    Map.keysSet establishedReady
+     <> (Set.fromList (PSQ.keys nextActivateTimes))
+     == Map.keysSet establishedStatus
+
+     -- there are ony warm peers in 'nextActiveTimes'
+  && all (== PeerWarm)
+         (Map.filterWithKey
+           (\peeraddr _ -> PSQ.member peeraddr nextActivateTimes)
+           establishedStatus)
+
+
+-- | Map of all established connections.
+--
+-- Complexity: /O(n*log n)/
+--
+toMap :: Ord peeraddr
```

Comment:
> The alternative would have been to say that the PSQ is always a subset of the established set. In that case it might be a OrdPSQ peeraddr Time () instead.

I think that would be confusing, after all the peers in the `PSQ` are never in established state (`Warm` or `Hot` state).

A better name would be `allPeers`.

Reply:
Would just name change be enough?

Reply:
Given how often we use toMap elsewhere, I think it'd make sense to change the invariant and make toMap/allPeers be O(1), with the less-frequently used ones requiring some extra computation.

Reply:
Agreed, I'll work on that.